### ISSUE TITLE
https does not work on localhost

### ISSUE THREAD

**Issue by kordeviant** [CrowdHailer/watercooler#6]
cloned the repo , started it , https://localhost:8443 not showing anything though server logs many things

**Comment by CrowdHailer** [comment]
What are the many things that are logged. can you copy them here

**Comment by kordeviant** [comment]
sorry for bad issue report: i am running on windows elixir 1.8
```
11:02:46.340 [error] GenServer #PID<0.1392.0> terminating
** (MatchError) no match of right hand side value: {:error, :closed}
    (ace) lib/ace/http/server.ex:119: Ace.HTTP.Server.handle_call/3
    (stdlib) gen_server.erl:661: :gen_server.try_handle_call/4
    (stdlib) gen_server.erl:690: :gen_server.handle_msg/6
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Last message (from #PID<0.1391.0>): {:accept, {:ssl, {:sslsocket, nil, {#Port<0.771>, {:config, {:ssl_options, :tls, [{3, 3}, {3, 2}, {3, 1}], :verify_none, {#Function<8.77706381/3 in :ssl.handle_verify_options/2>, []}, #Function<9.77706381/1 in :ssl.handle_verify_options/2>, false, false, :undefined, 1, "c:/code/elixir/watercooler/_build/dev/lib/watercooler/priv/localhost/certificate.pem", :undefined, "c:/code/elixir/watercooler/_build/dev/lib/watercooler/priv/localhost/certificate_key.pem", :undefined, [], :undefined, "", :undefined, :undefined, :undefined, :undefined, :undefined, [<<192, 44>>, <<192, 48>>, <<192, 36>>, <<192, 40>>, <<192, 46>>, <<192, 50>>, <<192, 38>>, <<192, 42>>, <<0, 159>>, <<0, 163>>, <<0, 107>>, <<0, 106>>, <<192, 43>>, <<192, 47>>, <<192, ...>>, <<...>>, ...], #Function<2.77706381/4 in :ssl.handle_options/3>, true, 268435456, true, true, :infinity, false, :undefined, ["h2", "http/1.1"], :undefined, :undefined, true, :undefined, [], :undefined, false, ...}, [send_timeout_close: true, send_timeout: 10000, reuseaddr: true, packet_size: 0, packet: 0, header: 0, active: false, mode: :binary], #PID<0.1373.0>, :undefined, [send_timeout_close: true, send_timeout: 10000, reuseaddr: true, packet_size: 0, packet: 0, header: 0, active: false, mode: :binary], {:gen_tcp, :tcp, :tcp_closed, :tcp_error}, :tls_connection}}}}}
State: %Ace.HTTP.Server{settings: [port: 8443, cleartext: false, certfile: "c:/code/elixir/watercooler/_build/dev/lib/watercooler/priv/localhost/certificate.pem",
keyfile: "c:/code/elixir/watercooler/_build/dev/lib/watercooler/priv/localhost/certificate_key.pem", port: 8080, cleartext: true], socket: nil, worker_supervisor:
#PID<0.1374.0>}
Client #PID<0.1391.0> is alive

    (stdlib) gen_server.erl:403: :gen_server.loop/7
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
iex(1)>
11:02:46.516 [debug] SERVER (port: 8443) sent: SETTINGS()
iex(1)>
11:02:46.516 [debug] SERVER (port: 8443) received: SETTINGS(header_table_size: 65536, max_concurrent_streams: 1000, initial_window_size: 6291456)
iex(1)>
11:02:46.516 [debug] SERVER (port: 8443) received: WINDOWN_UPDATE(stream_id: 0, increment: 15663105)

11:02:46.516 [debug] SERVER (port: 8443) sent: SETTINGS(ack: true)

11:02:46.516 [debug] SERVER (port: 8443) received: HEADERS(stream_id: 1, end_headers: true, end_stream: true)
iex(1)>
11:02:46.541 [info]  GET /
iex(1)>
11:02:46.541 [debug] SERVER (port: 8443) received: SETTINGS(ack: true)
iex(1)>
11:02:46.546 [info]  Sent 200 in 0┬╡s
iex(1)>
11:02:46.548 [error] GenServer #PID<0.1394.0> terminating
** (MatchError) no match of right hand side value: [[[[] | "<!DOCTYPE html>\r\n<html>\r\n  <head>\r\n    <meta charset=\"utf-8\">\r\n    <title>watercooler </title>\r\n    <link rel=\"stylesheet\" href=\"//fonts.googleapis.com/css?family=Roboto:300,300italic,700,700italic\">\r\n    <link rel=\"stylesheet\" href=\"/main.css\">\r\n  </head>\r\n  <body>\r\n  "], [[[[] | "<main>\r\n  <h1>Watercooler</h1>\r\n  <h2>Node: "], "nonode@nohost"] | "</h2>\r\n  <iframe name=\"myframe\" width=\"0\" height=\"0\"></iframe>\r\n  <form action=\"/publish\" method=\"post\" target=\"myframe\">\r\n    <input type=\"text\" name=\"message\">\r\n    <button type=\"submit\">Publish</button>\r\n  </form>\r\n  <hr>\r\n  <h1>Messages</h1>\r\n  <ul id=\"messages\"></ul>\r\n</main>\r\n"]] | "\r\n  <script type=\"text/javascript\" src=\"/main.js\"> </script>\r\n  </body>\r\n</html>\r\n"]
    (ace) lib/ace/http2/connection.ex:404: Ace.HTTP2.Connection.pack_data/4
    (ace) lib/ace/http2/connection.ex:369: Ace.HTTP2.Connection.pop_stream/3
    (ace) lib/ace/http2/connection.ex:247: anonymous fn/2 in Ace.HTTP2.Connection.send_available/1
    (stdlib) maps.erl:257: :maps.fold_1/3
    (ace) lib/ace/http2/connection.ex:177: anonymous fn/3 in Ace.HTTP2.Connection.handle_call/3
    (elixir) lib/enum.ex:1940: Enum."-reduce/3-lists^foldl/2-0-"/3
    (ace) lib/ace/http2/connection.ex:151: Ace.HTTP2.Connection.handle_call/3
    (stdlib) gen_server.erl:661: :gen_server.try_handle_call/4
Last message (from #PID<0.1598.0>): {:send, %Ace.HTTP.Channel{endpoint: #PID<0.1394.0>, id: 1, socket: :h2_socket}, [%Raxx.Response{body: [[[[] | "<!DOCTYPE html>\r\n<html>\r\n  <head>\r\n    <meta charset=\"utf-8\">\r\n    <title>watercooler </title>\r\n    <link rel=\"stylesheet\" href=\"//fonts.googleapis.com/css?family=Roboto:300,300italic,700,700italic\">\r\n    <link rel=\"stylesheet\" href=\"/main.css\">\r\n  </head>\r\n  <body>\r\n  "], [[[[] | "<main>\r\n  <h1>Watercooler</h1>\r\n  <h2>Node: "], "nonode@nohost"] | "</h2>\r\n  <iframe name=\"myframe\" width=\"0\" height=\"0\"></iframe>\r\n  <form action=\"/publish\" method=\"post\" target=\"myframe\">\r\n    <input type=\"text\" name=\"message\">\r\n    <button type=\"submit\">Publish</button>\r\n  </form>\r\n  <hr>\r\n  <h1>Messages</h1>\r\n  <ul id=\"messages\"></ul>\r\n</main>\r\n"]] | "\r\n  <script type=\"text/javascript\" src=\"/main.js\"> </script>\r\n  </body>\r\n</html>\r\n"], headers: [{"content-type", "text/html"}, {"content-length", "692"}], status: 200}]}
State: {"", %Ace.HTTP2.Connection{decode_context: {:hpack_context, {:dynamic_table, [{62, "accept-language", "en-US,en;q=0.9,fa;q=0.8"}, {63, "accept-encoding", "gzip, deflate, br"}, {64, "accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8"}, {65, "user-agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36"}, {66, "upgrade-insecure-requests", "1"}, {67, ":authority", "localhost:8443"}], 4096, 527}, 4096}, encode_context: {:hpack_context, {:dynamic_table, [], 4096, 0}, 4096}, local_settings: %Ace.HTTP2.Settings{enable_push: false, initial_window_size: 65535, max_concurrent_streams: 10000, max_frame_size: 16384}, max_peer_stream_id: 1, name: "SERVER (port: 8443)", next: :any, next_local_stream_id: 2, outbound_window: 15728640, peer: :server, pings: %{}, queued_settings: [], remote_settings: %Ace.HTTP2.Settings{enable_push: true, initial_window_size: 6291456, max_concurrent_streams: 1000, max_frame_size: 16384}, socket: {:ssl, {:sslsocket, {:gen_tcp, #Port<0.780>, :tls_connection, #PID<0.1373.0>}, #PID<0.1597.0>}}, stream_supervisor: #PID<0.1374.0>, streams: %{1 => %Ace.HTTP2.Stream{id: 1, incremented: 0, initial_window_size: 6291456, monitor: #Reference<0.2571311625.726401028.29937>, queue: [], sent: 0, status: {:idle, :closed}, worker: #PID<0.1598.0>}}}}
Client #PID<0.1598.0> is alive

    (stdlib) gen.erl:169: :gen.do_call/4
    (elixir) lib/gen_server.ex:986: GenServer.call/3
    (ace) lib/ace/http/channel.ex:48: Ace.HTTP.Channel.send/2
    (ace) lib/ace/http/worker.ex:82: Ace.HTTP.Worker.do_send/2
    (stdlib) gen_server.erl:637: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:711: :gen_server.handle_msg/6
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
iex(1)>
11:02:46.550 [error] GenServer #PID<0.1394.0> terminating
** (MatchError) no match of right hand side value: [[[[] | "<!DOCTYPE html>\r\n<html>\r\n  <head>\r\n    <meta charset=\"utf-8\">\r\n    <title>watercooler </title>\r\n    <link rel=\"stylesheet\" href=\"//fonts.googleapis.com/css?family=Roboto:300,300italic,700,700italic\">\r\n    <link rel=\"stylesheet\" href=\"/main.css\">\r\n  </head>\r\n  <body>\r\n  "], [[[[] | "<main>\r\n  <h1>Watercooler</h1>\r\n  <h2>Node: "], "nonode@nohost"] | "</h2>\r\n  <iframe name=\"myframe\" width=\"0\" height=\"0\"></iframe>\r\n  <form action=\"/publish\" method=\"post\" target=\"myframe\">\r\n    <input type=\"text\" name=\"message\">\r\n    <button type=\"submit\">Publish</button>\r\n  </form>\r\n  <hr>\r\n  <h1>Messages</h1>\r\n  <ul id=\"messages\"></ul>\r\n</main>\r\n"]] | "\r\n  <script type=\"text/javascript\" src=\"/main.js\"> </script>\r\n  </body>\r\n</html>\r\n"]
    (ace) lib/ace/http2/connection.ex:404: Ace.HTTP2.Connection.pack_data/4
    (ace) lib/ace/http2/connection.ex:369: Ace.HTTP2.Connection.pop_stream/3
    (ace) lib/ace/http2/connection.ex:247: anonymous fn/2 in Ace.HTTP2.Connection.send_available/1
    (stdlib) maps.erl:257: :maps.fold_1/3
    (ace) lib/ace/http2/connection.ex:177: anonymous fn/3 in Ace.HTTP2.Connection.handle_call/3
    (elixir) lib/enum.ex:1940: Enum."-reduce/3-lists^foldl/2-0-"/3
    (ace) lib/ace/http2/connection.ex:151: Ace.HTTP2.Connection.handle_call/3
    (stdlib) gen_server.erl:661: :gen_server.try_handle_call/4
Last message (from #PID<0.1393.0>): {:accept, {:ssl, {:sslsocket, nil, {#Port<0.771>, {:config, {:ssl_options, :tls, [{3, 3}, {3, 2}, {3, 1}], :verify_none, {#Function<8.77706381/3 in :ssl.handle_verify_options/2>, []}, #Function<9.77706381/1 in :ssl.handle_verify_options/2>, false, false, :undefined, 1, "c:/code/elixir/watercooler/_build/dev/lib/watercooler/priv/localhost/certificate.pem", :undefined, "c:/code/elixir/watercooler/_build/dev/lib/watercooler/priv/localhost/certificate_key.pem", :undefined, [], :undefined, "", :undefined, :undefined, :undefined, :undefined, :undefined, [<<192, 44>>, <<192, 48>>, <<192, 36>>, <<192, 40>>, <<192, 46>>, <<192, 50>>, <<192, 38>>, <<192, 42>>, <<0, 159>>, <<0, 163>>, <<0, 107>>, <<0, 106>>, <<192, 43>>, <<192, 47>>, <<192, ...>>, <<...>>, ...], #Function<2.77706381/4 in :ssl.handle_options/3>, true, 268435456, true, true, :infinity, false, :undefined, ["h2", "http/1.1"], :undefined, :undefined, true, :undefined, [], :undefined, false, ...}, [send_timeout_close: true, send_timeout: 10000, reuseaddr: true, packet_size: 0, packet: 0, header: 0, active: false, mode: :binary], #PID<0.1373.0>, :undefined, [send_timeout_close: true, send_timeout: 10000, reuseaddr: true, packet_size: 0, packet: 0, header: 0, active: false, mode: :binary], {:gen_tcp, :tcp, :tcp_closed, :tcp_error}, :tls_connection}}}}}
State: %Ace.HTTP.Server{settings: [port: 8443, cleartext: false, certfile: "c:/code/elixir/watercooler/_build/dev/lib/watercooler/priv/localhost/certificate.pem",
keyfile: "c:/code/elixir/watercooler/_build/dev/lib/watercooler/priv/localhost/certificate_key.pem", port: 8080, cleartext: true], socket: nil, worker_supervisor:
#PID<0.1374.0>}
Client #PID<0.1393.0> is alive

    (stdlib) gen_server.erl:403: :gen_server.loop/7
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
iex(1)>
11:02:46.553 [error] GenServer #PID<0.1598.0> terminating
** (stop) exited in: GenServer.call(#PID<0.1394.0>, {:send, %Ace.HTTP.Channel{endpoint: #PID<0.1394.0>, id: 1, socket: :h2_socket}, [%Raxx.Response{body: [[[[] | "<!DOCTYPE html>\r\n<html>\r\n  <head>\r\n    <meta charset=\"utf-8\">\r\n    <title>watercooler </title>\r\n    <link rel=\"stylesheet\" href=\"//fonts.googleapis.com/css?family=Roboto:300,300italic,700,700italic\">\r\n    <link rel=\"stylesheet\" href=\"/main.css\">\r\n  </head>\r\n  <body>\r\n  "], [[[[] | "<main>\r\n  <h1>Watercooler</h1>\r\n  <h2>Node: "], "nonode@nohost"] | "</h2>\r\n  <iframe name=\"myframe\" width=\"0\" height=\"0\"></iframe>\r\n  <form action=\"/publish\" method=\"post\" target=\"myframe\">\r\n    <input type=\"text\" name=\"message\">\r\n    <button type=\"submit\">Publish</button>\r\n  </form>\r\n  <hr>\r\n  <h1>Messages</h1>\r\n  <ul id=\"messages\"></ul>\r\n</main>\r\n"]] | "\r\n  <script type=\"text/javascript\" src=\"/main.js\"> </script>\r\n  </body>\r\n</html>\r\n"], headers: [{"content-type", "text/html"}, {"content-length", "692"}], status: 200}]}, 5000)
    ** (EXIT) an exception was raised:
        ** (MatchError) no match of right hand side value: [[[[] | "<!DOCTYPE html>\r\n<html>\r\n  <head>\r\n    <meta charset=\"utf-8\">\r\n    <title>watercooler </title>\r\n    <link rel=\"stylesheet\" href=\"//fonts.googleapis.com/css?family=Roboto:300,300italic,700,700italic\">\r\n    <link rel=\"stylesheet\" href=\"/main.css\">\r\n  </head>\r\n  <body>\r\n  "], [[[[] | "<main>\r\n  <h1>Watercooler</h1>\r\n  <h2>Node: "], "nonode@nohost"] | "</h2>\r\n  <iframe name=\"myframe\" width=\"0\" height=\"0\"></iframe>\r\n  <form action=\"/publish\" method=\"post\" target=\"myframe\">\r\n    <input type=\"text\" name=\"message\">\r\n    <button type=\"submit\">Publish</button>\r\n  </form>\r\n  <hr>\r\n  <h1>Messages</h1>\r\n  <ul id=\"messages\"></ul>\r\n</main>\r\n"]] | "\r\n  <script type=\"text/javascript\" src=\"/main.js\"> </script>\r\n  </body>\r\n</html>\r\n"]
            (ace) lib/ace/http2/connection.ex:404: Ace.HTTP2.Connection.pack_data/4
            (ace) lib/ace/http2/connection.ex:369: Ace.HTTP2.Connection.pop_stream/3
            (ace) lib/ace/http2/connection.ex:247: anonymous fn/2 in Ace.HTTP2.Connection.send_available/1
            (stdlib) maps.erl:257: :maps.fold_1/3
            (ace) lib/ace/http2/connection.ex:177: anonymous fn/3 in Ace.HTTP2.Connection.handle_call/3
            (elixir) lib/enum.ex:1940: Enum."-reduce/3-lists^foldl/2-0-"/3
            (ace) lib/ace/http2/connection.ex:151: Ace.HTTP2.Connection.handle_call/3
            (stdlib) gen_server.erl:661: :gen_server.try_handle_call/4
    (elixir) lib/gen_server.ex:989: GenServer.call/3
    (ace) lib/ace/http/channel.ex:48: Ace.HTTP.Channel.send/2
    (ace) lib/ace/http/worker.ex:82: Ace.HTTP.Worker.do_send/2
    (stdlib) gen_server.erl:637: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:711: :gen_server.handle_msg/6
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Last message: {%Ace.HTTP.Channel{endpoint: #PID<0.1394.0>, id: 1, socket: :h2_socket}, %Raxx.Request{authority: "localhost:8443", body: false, headers: [{"upgrade-insecure-requests", "1"}, {"user-agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36"}, {"accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8"}, {"accept-encoding", "gzip, deflate, br"}, {"accept-language", "en-US,en;q=0.9,fa;q=0.8"}], method: :GET, mount: [], path: [], query: nil, raw_path: "/", scheme: :https}}
State: %Ace.HTTP.Worker{app_module: Watercooler.WWW, app_state: %{}, channel: %Ace.HTTP.Channel{endpoint: #PID<0.1394.0>, id: 1, socket: :h2_socket}, channel_monitor: #Reference<0.2571311625.726401028.29936>}
```

**Comment by CrowdHailer** [comment]
So, I have seen this before. The key line is.
```
(MatchError) no match of right hand side value: [[[[] | "<!DOCTYPE html>\r\n<h...
    (ace) lib/ace/http2/connection.ex:404: Ace.HTTP2.Connection.pack_data/4
```

`Raxx.View` now returns an iolist in the response when using render.
When using https the HTTP/2 connection is used. Unfortunately the HTTP/2 implementation does not handle iolists properly yet. So this is really a deficiency in Ace.

To double check if I have diagnosed this correctly, if you add the following to the last line of the controller does it fix it.
```elixir
response = response(:ok) |>  # etc
%{response | body: :erlang.iolist_to_binary(response.body)}
```

**Comment by CrowdHailer** [comment]
Fixed by updating Ace to 0.18.3 https://github.com/CrowdHailer/watercooler/commit/1b5e0f3d91340ac02fc75d78995d71c5e20973b7

Use latest master of this repo to include Ace 0.1.8.3

**Comment by kordeviant** [comment]
thanks, sorry that I did not get to test your suggestion yet.
